Mathlib/Algebra/BigOperators/Fin.lean

```diff
@@ -0,0 +1,383 @@
+constant function. These results have variants for sums instead of products.
+
+-/
+
+-- Porting note: Uncomment once localised BigOperators is fixed:
+-- open BigOperators
+
+open Finset
+
+variable {α : Type _} {β : Type _}
+
+namespace Finset
+
+@[to_additive]
+theorem prod_range [CommMonoid β] {n : ℕ} (f : ℕ → β) :
+    (∏ i in Finset.range n, f i) = ∏ i : Fin n, f i :=
+  prod_bij' (fun k w => ⟨k, mem_range.mp w⟩) (fun _ _ => mem_univ _)
+    (fun _ _ => congr_arg _ (Fin.val_mk _).symm) (fun a _ => a) (fun a _ => mem_range.mpr a.prop)
+    (fun _ _ => Fin.val_mk _) fun _ _ => Fin.eta _ _
+#align finset.prod_range Finset.prod_range
+#align finset.sum_range Finset.sum_range
+
+end Finset
+
+namespace Fin
+
+@[to_additive]
+theorem prod_univ_def [CommMonoid β] {n : ℕ} (f : Fin n → β) :
+    (∏ i, f i) = ((List.finRange n).map f).prod := by simp [univ_def]
+#align fin.prod_univ_def Fin.prod_univ_def
+#align fin.sum_univ_def Fin.sum_univ_def
+
+@[to_additive]
+theorem prod_ofFn [CommMonoid β] {n : ℕ} (f : Fin n → β) : (List.ofFn f).prod = ∏ i, f i := by
+  rw [List.ofFn_eq_map, prod_univ_def]
+#align fin.prod_of_fn Fin.prod_ofFn
+#align fin.sum_of_fn Fin.sum_ofFn
+
+/-- A product of a function `f : fin 0 → β` is `1` because `fin 0` is empty -/
+@[to_additive "A sum of a function `f : fin 0 → β` is `0` because `fin 0` is empty"]
+theorem prod_univ_zero [CommMonoid β] (f : Fin 0 → β) : (∏ i, f i) = 1 :=
+  rfl
+#align fin.prod_univ_zero Fin.prod_univ_zero
+#align fin.sum_univ_zero Fin.sum_univ_zero
+
+/-- A product of a function `f : fin (n + 1) → β` over all `fin (n + 1)`
+is the product of `f x`, for some `x : fin (n + 1)` times the remaining product -/
+@[to_additive "A sum of a function `f : fin (n + 1) → β` over all `fin (n + 1)` is the sum of
+`f x`, for some `x : fin (n + 1)` plus the remaining product"]
+theorem prod_univ_succAbove [CommMonoid β] {n : ℕ} (f : Fin (n + 1) → β) (x : Fin (n + 1)) :
+    (∏ i, f i) = f x * ∏ i : Fin n, f (x.succAbove i) := by
+  rw [univ_succAbove, prod_cons, Finset.prod_map]
+#align fin.prod_univ_succ_above Fin.prod_univ_succAbove
+#align fin.sum_univ_succ_above Fin.sum_univ_succAbove
+
+/-- A product of a function `f : fin (n + 1) → β` over all `fin (n + 1)`
+is the product of `f 0` plus the remaining product -/
+@[to_additive "A sum of a function `f : fin (n + 1) → β` over all `fin (n + 1)` is the sum of
+`f 0` plus the remaining product"]
+theorem prod_univ_succ [CommMonoid β] {n : ℕ} (f : Fin (n + 1) → β) :
+    (∏ i, f i) = f 0 * ∏ i : Fin n, f i.succ :=
+  prod_univ_succAbove f 0
+#align fin.prod_univ_succ Fin.prod_univ_succ
+#align fin.sum_univ_succ Fin.sum_univ_succ
+
+/-- A product of a function `f : fin (n + 1) → β` over all `fin (n + 1)`
+is the product of `f (fin.last n)` plus the remaining product -/
+@[to_additive "A sum of a function `f : fin (n + 1) → β` over all `fin (n + 1)` is the sum of
+`f (fin.last n)` plus the remaining sum"]
```

Comment:
```suggestion
/-- A product of a function `f : Fin 0 → β` is `1` because `Fin 0` is empty -/
@[to_additive "A sum of a function `f : Fin 0 → β` is `0` because `Fin 0` is empty"]
theorem prod_univ_zero [CommMonoid β] (f : Fin 0 → β) : (∏ i, f i) = 1 :=
  rfl
#align fin.prod_univ_zero Fin.prod_univ_zero
#align fin.sum_univ_zero Fin.sum_univ_zero

/-- A product of a function `f : Fin (n + 1) → β` over all `Fin (n + 1)`
is the product of `f x`, for some `x : Fin (n + 1)` times the remaining product -/
@[to_additive "A sum of a function `f : Fin (n + 1) → β` over all `Fin (n + 1)` is the sum of
`f x`, for some `x : Fin (n + 1)` plus the remaining product"]
theorem prod_univ_succAbove [CommMonoid β] {n : ℕ} (f : Fin (n + 1) → β) (x : Fin (n + 1)) :
    (∏ i, f i) = f x * ∏ i : Fin n, f (x.succAbove i) := by
  rw [univ_succAbove, prod_cons, Finset.prod_map]
#align fin.prod_univ_succ_above Fin.prod_univ_succAbove
#align fin.sum_univ_succ_above Fin.sum_univ_succAbove

/-- A product of a function `f : Fin (n + 1) → β` over all `Fin (n + 1)`
is the product of `f 0` plus the remaining product -/
@[to_additive "A sum of a function `f : Fin (n + 1) → β` over all `Fin (n + 1)` is the sum of
`f 0` plus the remaining product"]
theorem prod_univ_succ [CommMonoid β] {n : ℕ} (f : Fin (n + 1) → β) :
    (∏ i, f i) = f 0 * ∏ i : Fin n, f i.succ :=
  prod_univ_succAbove f 0
#align fin.prod_univ_succ Fin.prod_univ_succ
#align fin.sum_univ_succ Fin.sum_univ_succ

/-- A product of a function `f : Fin (n + 1) → β` over all `Fin (n + 1)`
is the product of `f (Fin.last n)` plus the remaining product -/
@[to_additive "A sum of a function `f : Fin (n + 1) → β` over all `Fin (n + 1)` is the sum of
`f (Fin.last n)` plus the remaining sum"]
```